src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java

```diff
@@ -350,6 +350,7 @@ private RepositoryFunction getHandler(Rule rule) {
       // Upon an exceptional exit, the fetching of that repository is over as well.
       env.getListener().post(new RepositoryFetching(repositoryName, true));
       env.getListener().post(new RepositoryFailedEvent(repositoryName, e.getMessage()));
+      env.getListener().handle(Event.error("Error fetching repository: " + e.getMessage()));
       throw e;
```

Comment:
Your change doesn't make the situation worse, but it's going to cause further problems that this SkyFunction is throwing random exceptions underneath RepositoryFunctionException (which is unwrapped anyway). It would make more sense to me if the underlying exception were wrapped in a RepositoryException or similar. That would let callers know that they shouldn't try to handle or show errors coming from here. As it stands, some callers might try to do something with the IOException.

Reply:
I'm generally of the opinion that low level exceptions (like IOException) should not be known to reverse deps and should instead be wrapped in some local exception. This would allow a cleaner interface between SkyFunctions.

I thought about doing that for `RepositoryFunctionException` but was discouraged by the fact that a few callers of `RepositoryDelegatorFunction` are handling `IOException`. We now have a mix of both styles: some functions like `ActionExecutionFunction` repackages `IOException` as `ActionExecutionException`, while some like `RepositoryFunctionException` basically expects callers to know all the underlying exceptions that could be thrown.

> That would let callers know that they shouldn't try to handle or show errors coming from here. As it stands, some callers might try to do something with the IOException.

Sure, as long as we document somewhere that one shouldn't catch and re-report `RepositoryException`.

Reply:
I don't see any place where RepositoryDirectoryValue#key is passed to env.getValueOrThrow. Do you just mean that those SkyFunctions separately handle IOExceptions? That makes sense: if you have to directly declare a dep on a file, you should get an IOException.

As I said before, I think the principle is that you should enrich your exception as much as you are able to. With RepositoryDelegatorFunction, it seems to be fully enriched, so we should just report and encapsulate it as a specific type.

Documenting that you shouldn't catch RepositoryException in RepositoryException's javadoc seems straightforward.

Reply:
>  but was discouraged by the fact that a few callers of RepositoryDelegatorFunction are handling IOException.

Sorry, my memory from last Friday failed me... It's not that "a few callers of RepositoryDelegatorFunction are handling IOException", but rather there are 6 implementations of [`RepositoryFunction#fetch`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryFunction.java;l=132;drc=0617f2c7b3f442a00f7359d216b8fab14d2dd762) (which originates the `RepositoryFunctionException`), each throwing `RepositoryFunctionException` with a different underlying exception. 

As a result, it's hairier to encapsulate those in a `RepositoryException` before wrapping that in `RepositoryFunctionException`, but it's not infeasible. Was that what you had in mind?

Reply:
Changes:
- Wrapped the low level exceptions (IOException, ...) into an AlreadyReportedRepositoryAccessException before wrapping that in RepositoryFunctionException.
- Fixed the failing tests (mostly because of the failFastHandler).